scanner/cmd/updater/main.go

```diff
@@ -2,25 +2,54 @@ package main
 
 import (
 	"context"
+	"errors"
 	"flag"
 	"log"
+	"time"
 
 	"github.com/quay/zlog"
 	"github.com/stackrox/rox/scanner/updater"
 )
 
+func tryExport(outputDir string) error {
+	const timeout = 35 * time.Minute
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	err := updater.Export(ctx, outputDir)
+	if errors.Is(err, context.DeadlineExceeded) {
+		zlog.Error(ctx).Err(err).Msg("Export timed out")
+		return err
+	}
+	if err != nil {
+		zlog.Error(ctx).Err(err).Msg("Failed to export the vulnerabilities")
+		return err
+	}
+
+	return nil
+}
+
 func main() {
-	// Parse command-line flags
 	outputDir := flag.String("output-dir", "", "Output directory")
 	flag.Parse()
 
-	// Check if outputDir flag is provided
 	if *outputDir == "" {
 		log.Fatal("Missing argument for the output directory.")
 	}
 
-	ctx := context.Background()
-	if err := updater.Export(ctx, *outputDir); err != nil {
-		zlog.Error(ctx).Err(err).Send()
+	const maxRetries = 3
+
+	for attempt := 1; attempt <= maxRetries; attempt++ {
+		err := tryExport(*outputDir)
+		if err == nil {
+			break
```

Comment:
I just realized we exit normally if we exceed the number of retries, which we shouldn't do.

I think we should return here and log an error outside the loop, as we would not expect to reach there. What do you think?

Reply:
I agree. But with this logic, `log.fatal` will be executed if any error (except for timeout) happens in one of the three attempts. Do we still want to log outside of the loop? 

```
for attempt := 1; attempt <= maxRetries; attempt++ {
		err := tryExport(*outputDir)
		if err == nil {
			return
		}
		if errors.Is(err, context.DeadlineExceeded) {
			zlog.Error(context.Background()).Err(err).Msg("Data export attempt failed; will attempt retry if within retry limits")
			continue
		}

		zlog.Error(context.Background()).Err(err).Msg("Data updater failed to update vulnerabilities")
		log.Fatal("The data export process has failed.")
	}
```

Reply:
Yes, but let's say we have three timeouts. That should be a failure, right? If we don't `log.Fatal` outside the loop (assuming you keep the `return` when `err == nil`), then it'll have a passing exit status even though we were unable to get the vuln data, right?